src/db_adapters/access_keys.rs

```diff
@@ -0,0 +1,141 @@
+                _ => false,
+            }
+        })
+        .filter_map(|outcome_with_receipt| outcome_with_receipt.receipt.as_ref());
+
+    let mut access_keys = HashMap::<(String, String), models::access_keys::AccessKey>::new();
+
+    for receipt in successful_receipts {
+        if let near_primitives::views::ReceiptEnumView::Action { actions, .. } = &receipt.receipt {
+            for action in actions {
+                match action {
+                    near_primitives::views::ActionView::AddKey {
+                        public_key,
+                        access_key,
+                    } => {
+                        if let Some(existing_record) = access_keys
+                            .get_mut(&(public_key.to_string(), receipt.receiver_id.to_string()))
+                        {
+                            existing_record.created_by_receipt_id =
+                                Some(receipt.receipt_id.to_string());
+                            existing_record.deleted_by_receipt_id = None;
+                        } else {
+                            access_keys.insert(
+                                (public_key.to_string(), receipt.receiver_id.to_string()),
+                                models::access_keys::AccessKey::from_action_view(
+                                    public_key,
+                                    &receipt.receiver_id,
+                                    access_key,
+                                    &receipt.receipt_id,
+                                ),
+                            );
+                        }
+                    }
+                    near_primitives::views::ActionView::DeleteKey { public_key } => {
+                        if let Some(existing_record) = access_keys
+                            .get_mut(&(public_key.to_string(), receipt.receiver_id.to_string()))
+                        {
+                            existing_record.deleted_by_receipt_id =
+                                Some(receipt.receipt_id.to_string());
+                        } else {
+                            access_keys.insert(
+                                (public_key.to_string(), receipt.receiver_id.to_string()),
+                                models::access_keys::AccessKey {
+                                    public_key: public_key.to_string(),
+                                    account_id: receipt.receiver_id.to_string(),
+                                    created_by_receipt_id: None,
+                                    deleted_by_receipt_id: Some(receipt.receipt_id.to_string()),
+                                    permission: models::enums::AccessKeyPermission::NotApplicable,
+                                },
+                            );
+                        }
+                    }
+                    _ => continue,
+                }
+            }
+        }
+    }
+
+    let mut access_keys_to_insert: Vec<models::access_keys::AccessKey> = vec![];
```

Comment:
You can [`partition`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.partition) the `access_keys`, and run the queries in parallel (using FuturesUnordered for the updates + async block + `join!(updates_block, inserts_block)`)